src/ucsi/ppm/state_machine.rs

```diff
@@ -0,0 +1,191 @@
+//! This module defines the core PPM state machine as defined in 6.1 of the UCSI spec.
+//! The state machine presented in the spec is a combination of state (circular boxes) as well as
+//! actions (rectangular boxes) that should be done when transitioning between states. How to perform these actions
+//! is left to the implementation. The state machine presented here abstracts over these actions with the [`Input`] enum.
+//! The [`Output`] enum defines the notifications to the OPM.
+
+/// PPM states
+///
+/// The spec does not explicitly distinguish between the BusyNotificationDisabled and Busy states. But this is done here
+/// because otherwise it would allow to transition from the IdleNotificationDisabled state to the Idle state without
+/// notifications being enabled.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum State {
+    /// Idle, notification disabled
+    IdleNotificationDisabled,
+    /// Busy
+    BusyNotificationDisabled,
+    /// Idle (notification enabled)
+    Idle,
+    /// Busy (notification enabled)
+    Busy,
+    /// Processing Command,
+    ProcessingCommand,
+    /// Waiting for command complete ack
+    WaitForCommandCompleteAck,
+    /// Waiting for async event ack
+    WaitForAsyncEventAck,
+}
+
+impl State {
+    /// Returns true if the state is busy
+    pub fn is_busy(&self) -> bool {
+        matches!(self, State::BusyNotificationDisabled | State::Busy)
+    }
+}
+
+/// Inputs to the PPM state machine
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Input {
+    /// Notification enabled
+    NotificationEnabled,
+    /// PPM reset
+    Reset,
+    /// Current command cancelled
+    Cancel,
+    /// Command completed immediately
+    CommandImmediate,
+    /// Command did not complete immediately
+    CommandAsync,
+    /// Command completed,
+    CommandCompleted,
+    /// Command complete ack received from OPM
+    CommandCompleteAck,
+    /// Async event ack received from OPM
+    AsyncEventAck,
+    /// External busy status changed
+    BusyChanged,
+    /// There's a pending async event
+    PendingAsyncEvent,
+}
+
+/// Outputs from the PPM state machine
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Output {
+    /// Notify OPM that command completed
+    OpmNotifyCommandComplete,
+    /// Notify OPM of async event
+    OpmNotifyAsyncEvent,
+    /// Notify OPM of PPM reset
+    OpmNotifyReset,
+    /// Notify OPM that PPM is busy
+    OpmNotifyBusy,
+}
+
+/// Attempted transition that is not allowed by the state machine
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct InvalidTransition;
```

Comment:
Would it be worth returning more information in `InvalidTransition` such as the current state, valid inputs for that state, etc.?

Of note: my suggestion to use an input enum per state (or something similar) would naturally eliminate this error; there'd be no way to express an invalid transition in the code.